### PR TITLE
[rfc1213] Interface MIB add l3 vlan interfaces & aggregate rif counters

### PR DESCRIPTION
**Pull request by stepanblyschak**
**- What I did**

Rework based on https://github.com/Azure/sonic-snmpagent/pull/133 and feedback https://github.com/Azure/sonic-snmpagent/issues/148.

**- How I did it**

- merge #133 with upstream master.
- Fix UnavailableDataError when RIF counters are not enabled.
- Make RIF keys in maps unique in multi asic env.

**- How to verify it**

- Disable rif counters
- restart swss
- snmpwalk
- Verify no UnavailableDataError in logs
- Verify port counters are returned
- Enable rif counters
- snmpwalk
- Verify no error in logs
- Verify error in/out counters are aggregated for rif ports and portchannels
- Verify VLAN RIF counters are present in the MIB.

**- Description for the changelog**

Added support for aggregated router interface counters and L3 VLAN counters to RFC1213 MIB.

### CODE REVIEW

**Comment by qiluo-msft** [comment]
Please resolve the conflicts.

File: src/sonic_ax_impl/mibs/__init__.py [hunk 1]
```diff
@@ -43,20 +43,25 @@
...
 redis_kwargs = {'unix_socket_path': '/var/run/redis/redis.sock'}
 
-
-def get_neigh_info(neigh_key):
-    """
-    split neigh_key string of the format:
-    NEIGH_TABLE:device:ipv4_address
-    """
-    _, device, ip = neigh_key.split(':')
-    return device, ip
```
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/__init__.py]: >b' [](start = 11, length = 2)

If you merge latest master, you don't need to use byte array and just use string.

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py [hunk 2]
```diff
@@ -219,24 +194,40 @@ def reinit_data(self):
         self.mgmt_oid_name_map, \
         self.mgmt_alias_map = mibs.init_mgmt_interface_tables(self.db_conn[0])
 
+        self.vlan_name_map, \
```
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/ietf/rfc1213.py:231]: >bytes [](start = 29, length = 5)

If you merge latest master, you don't need to encoding to byte array.

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 from bisect import bisect_right
 
 from sonic_ax_impl import mibs
+from sonic_ax_impl import logger
```
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/ietf/rfc1213.py:7]: >logger [](start = 26, length = 6)

Merge with above line

**Review by qiluo-msft — COMMENTED** [review]
As comments

**Review by SuvarnaMeenakshi — COMMENTED** [review]
similar to test_interfaces, can you add changes in tests/namespace/test_interfaces.
Also, mock_db changes for multi-asic mock-dbs as well.

**Comment by qiluo-msft** [comment]
Please resolve LGTM alert. Others look good to me.

**Review by qiluo-msft — APPROVED** [review]
Please also check with other reviewers.

File: tests/namespace/test_interfaces.py [hunk 4]
```diff
@@ -237,7 +237,7 @@ def test_if_type_portchannel(self):
         self.assertEqual(value0.data, 161)
 
     def test_getnextpdu_first_bp_ifindex(self):
-        oid = ObjectIdentifier(11, 0, 0, 0, (1, 3, 6, 1, 2, 1, 2, 2, 1, 1,1004))
+        oid = ObjectIdentifier(11, 0, 0, 0, (1, 3, 6, 1, 2, 1, 2, 2, 1, 1, 8999))
```
Comment (SuvarnaMeenakshi) [inline tests/namespace/test_interfaces.py:240]: Is this change required? There is no new interface added in namespace mock_dbs, is that right?

**Comment by SuvarnaMeenakshi** [comment]
> similar to test_interfaces, can you add changes in tests/namespace/test_interfaces.
> Also, mock_db changes for multi-asic mock-dbs as well.

Closed.

File: tests/namespace/test_interfaces.py (hunk 4 above)
Comment (qiluo-msft) [reply tests/namespace/test_interfaces.py:240]: @stepanblyschak This comment is still applicable.

---
In reply to: [527836254](https://github.com/Azure/sonic-snmpagent/pull/169#discussion_r527836254) [](ancestors = 527836254)

**Comment by liat-grozovik** [comment]
@qiluo-msft can you please merge so we can proceed with the submodule update?

File: tests/namespace/test_interfaces.py (hunk 4 above)
Comment (Blyschak) [reply tests/namespace/test_interfaces.py:240]: I have added Ethernet16 https://github.com/Azure/sonic-snmpagent/pull/169/files/a00e69e25dfc7c17f2554a625b57c447aff477c0#diff-fe4c39c260164f04bfe84f3d68d6f6a5a28850a840667a26aaba67a9fc1afacbR48.
Now, different asic dbs have different port table.

Comment (stepanblyschak) [reply tests/namespace/test_interfaces.py:240]: I have added Ethernet16 https://github.com/Azure/sonic-snmpagent/pull/169/files/a00e69e25dfc7c17f2554a625b57c447aff477c0#diff-fe4c39c260164f04bfe84f3d68d6f6a5a28850a840667a26aaba67a9fc1afacbR48.
Now, different asic dbs have different port table.

**Review by SuvarnaMeenakshi — APPROVED** [review]

File: src/sonic_ax_impl/mibs/ietf/rfc1213.py [hunk 5]
```diff
@@ -302,6 +336,31 @@ def _get_counter(self, oid, table_name):
...
+        For ports with l3 router interfaces l3 drops may be counted separately (RIF counters)
+        add l3 drops to l2 drop counters cache according to mapping
+
+        For l3vlan map l3 counters to l2 counters
+        """
+        for rif_sai_id, port_sai_id in self.rif_port_map.items():
+            if port_sai_id in self.if_id_map:
+                port_idx = mibs.get_index_from_str(self.if_id_map[port_sai_id])
+                for port_counter_name, rif_counter_name in mibs.RIF_DROPS_AGGR_MAP.items():
+                    self.if_counters[port_idx][port_counter_name] = \
```
Comment (qiluo-msft) [inline src/sonic_ax_impl/mibs/ietf/rfc1213.py:350]: >self.if_counters [](start = 20, length = 16)

I see you change `self.if_counters[][]` type from `str` to `int`.
Although this is working in python, however we are moving the library to sonic-swss-common SWIG wrapper, and this will make SWIG confused.
Please make type fixed.
